cpp/src/arrow/util/iterator.h

```diff
@@ -138,11 +139,11 @@ class Iterator : public util::EqualityComparable<Iterator<T>> {
         value_ = IterationTraits<T>::End();
         return;
       }
-      value_ = iterator_.Next();
+      value_ = iterator_->Next();
     }
 
     Result<T> value_;
-    Iterator iterator_;
+    std::shared_ptr<Iterator> iterator_;
```

Comment:
This allows `Iterator::RangeIterator` to be copied, which is a requirement on iterators https://en.cppreference.com/w/cpp/named_req/Iterator

We haven't caught this before because we only use `Iterator::begin` and `Iterator::end` in the context of range-for loops which don't require it. However cython wraps the range-for loop does, however, which is why I noticed the concept error.